Max/Min/SD/Mean/IQR Ops

ilan-gold opened the pull request:
I am not sure what direction this should go with the large resolutions imagery, but I think this would really help with our static image data.  Do we want to use an "auto-adjust sliders" feature based on this data  in the demo?  We could do it just for the static image data.  I originally implemented it as +- 2 standard deviations from the mean but then removed it - it worked well across all of our images except the COVID-19 one, which I checked and has a very odd distribution with a large spike of junk data. This is just a draft - would like feedback.

manzt commented:
Basically I think determining min and max should be be an application-dependent feature and not a part of viv. Ultimately the users would need to interact with the loader manually to set sliders ranges regardless (in the UI framework). I think a good solution would be to add the `tfjs` stuff to the `demo` and then we can replicate this else where (like in vitessce and the portal).

ilan-gold commented:
Perhaps then we can export just a function like `getChannelStats({data})` or more succintly `getChannelStats({loader})`.  That would solve our bundling issue.  We don't ship with any modules except GeoTIFF which is needed for the decoding inline worker.  I think this seems like it'll be a fairly common operation that people will want to use with Viv, including collaborators at Indiana who are not using Vitessce or other future users we are aware of. Perhaps I should think on this more.

manzt commented:
I think it would be good to get input from users before adding this to viv. I'm not saying "no way" but it does feel like feature creep to me. 

I do think it's useful, but more of a cool application _with_ viv rather than an application _of_ viv. TFJS isn't the only library that can compute stats on typed arrays, and ultimately I'd hesitate to prescribe a solution.

ilan-gold commented:
This is another library I found: http://gpu.rocks/.  I just went with `tfjs` because it is what I have used before for GPGPU.  I think we do need to use GPGPU for the purpose of scalability - this is from the gpu.rocks website:

<img width="1341" alt="Screen Shot 2020-05-19 at 4 10 36 PM" src="https://user-images.githubusercontent.com/43999641/82373240-4c9ddd80-99eb-11ea-8ced-49574c4e5018.png">

For mass-spec data, CPU is fine but for CODEX raster imagery, which is ~1300 x 1100, it would not be great.  Even a 512x512 is 6 times faster.

manzt commented:
Right, I think this will be something that is useful for our web applications, but we can roll our own solution there (since it really is just a _thin_ wrapper around `tfjs`). That's different from a library export, which we will continue to need to maintain for others.

I'm not arguing whether `tfjs` is the right library for this particular use case in our applications, but I am saying that it should be decided by the application and not provided by viv. I think it would be useful to get input from others on our team ( @ngehlenborg) as others using viv before adding something like this. Perhaps better defining what are `viv`'s responsibilities because it seems like we disagree somewhat.

ilan-gold commented:
Hmm...did some benchmarking and the CPU looks faster.  I had removed the tfjs code previously in the process of this PR because I had a suspicion this might be the case.  I guess the question is whether or  not we want to bundle that solution.  The thing is, `Math.min` does not work for that.  You have to use an actual for-loop.  We could bundle this or at least leave a note for people.

ilan-gold commented:
We could also just do a while loop:
```javascript
while (len--) {
      if (arr[len] > max) {
        max = arr[len];
      }
    }
```

manzt commented:
Yes. My guess is that's due to tensorflow making copies of that data w/ GPU. I think there are some applications where it might make sense to have tfjs as a dependency but many where it is not. Agreed, there are many solutions for how to compute these stats, but I just don't think we should be the ones deciding that for another developer (who can definitely write a for-loop or use tfjs) if they find a use case where they need these values.

ilan-gold commented:
I guess my thing is that no matter (CPU or GPU) how we did this math/min/mean/sd op, it was going to be non-standard so I thought it should live here to prevent users from having to figure it out on their own.  I am fine not having this in Viv for now, but I still feel like the fact that someone would have to do research to figure this out feels odd given that it's something people will want to do commonly.  Perhaps this PR will serve as the "documentation" for people.  I'll add an "ARCHIVE" to the front of it.  Thanks for the feedback,

manzt commented:
Sorry to push back so much, but more than anything I think it would be valuable to talk to viv-users for whether they think think would be useful. It just feels like a decision for application-land, since applying the output of `getStats` is applied to UI components on the screen.

Pre-computing the min-max will always be most performant in terms of user experience.

ilan-gold commented:
No worries! I wouldn't have asked for feedback if I didn't want it!  We're on the cutting-edge here on some level and working in near-isolation so this stuff is tough.  We'll figure it out.  We can always put it in Vitessce - my main concern here is the CODEX data.

ilan-gold commented:
I updated the median calculation and added first and third quartiles.  Perhaps we need something stronger, or an interface do to arbitrary sections of the data (i.e quintiles or something else).  Using quickselect is also substantially faster than just sorting and indexing:

<img width="1662" alt="Screen Shot 2020-05-29 at 3 47 24 PM" src="https://user-images.githubusercontent.com/43999641/83299364-05ba9f80-a1c4-11ea-9845-b23156247ea8.png">

manzt commented on package.json:
```diff
@@ -106,6 +106,7 @@
     "fast-xml-parser": "^3.16.0",
     "geotiff": "^1.0.0-beta.11",
+    "quickselect": "^2.0.0",
     "zarr": "^0.2.3"
```
Need to make sure the latest version of zarr. I just released `0.3.0` (not beta) so please update to that!

ilan-gold replied:
Sorry, yes, just. noticed the merge conflict.

manzt commented on src/loaders/utils.js:
```diff
@@ -102,3 +104,75 @@ export function isBioformatsNoPadHeightVersion(software) {
+    let total = 0;
+    // Range (min/max).
+    // eslint-disable-next-line no-plusplus
+    while (len--) {
+      if (arr[len] < min) {
+        min = arr[len];
+      }
+      if (arr[len] > max) {
+        max = arr[len];
+      }
+      total += arr[len];
+    }
```
any reason not to use `for ... of`?

```suggestion
    for (let val of arr) {
      if (val < min) min = val;
      if (val > max) max = val;
      total += val
    }
```

ilan-gold replied:
Big performance gains to the while loop: https://jsbench.me/rtkawo35dg/1
<img width="1660" alt="Screen Shot 2020-06-01 at 11 49 29 AM" src="https://user-images.githubusercontent.com/43999641/83427014-0cccf200-a3fe-11ea-8eec-6693669d0eed.png">

manzt commented on src/loaders/utils.js:
```diff
@@ -102,3 +104,75 @@ export function isBioformatsNoPadHeightVersion(software) {
+    standardDeviations: [],
+    data,
+    firstQuartiles: [],
+    thirdQuartiles: []
```
Why include `data` in stats? I think having an array of `stats` objects would be a better data structure for this (compared to an object of arrays). The UI components use channel objects in Vitessce and then serialize to arrays only for rendering purposes with viv.

You have 4 channels, return the stats as objects for those channels. If an end user wanted stats for channel `i` accessing those would look like:

```javascript
const stats = {
  mean: channelStats.means[i],
  median: channelStats.medians[i],
  dataRange: channelStats.dataRanges[i],
  standardDeviation: channelStats.standardDeviations[i],
  firstQuartile: channelStats.firstQuartiles[i],
  thirdQuartile: channelStats.thirdQuartiles[i],
}
```

vs:

```javascript
const stats = channelStats[i];
const { mean, median, dataRange, standardDeviation, firstQuartile, thirdQuartile } = stats;
```

The former is much more error prone.

manzt replied:
nit, but some more convenient / common keys might be useful:

- `dataRange` --> `domain` (https://github.com/d3/d3-scale#continuous_domain)
- `standardDeviation` --> `sd`
- `firstQuartile` --> `q1`
- `thirdQuartile` --> `q3`

ilan-gold replied:
You're right about the data structure.  Will do.

ilan-gold replied:
As for `data`, it will be helpful for generating visualizations of the data along with stats, like a histogram.

manzt commented on src/loaders/utils.js:
```diff
@@ -102,3 +104,75 @@ export function isBioformatsNoPadHeightVersion(software) {
+    len = arr.length;
+    let sumSquared = 0;
+    // eslint-disable-next-line no-plusplus
+    while (len--) {
+      sumSquared += (arr[len] - mean) ** 2;
+    }
```
same as above

manzt commented on src/loaders/utils.js:
```diff
@@ -102,3 +104,75 @@ export function isBioformatsNoPadHeightVersion(software) {
+
+    // Median, and quartiles via quickselect: https://en.wikipedia.org/wiki/Quickselect.
+    // Odd number lengths should round down the index.
+    const mid = Math.floor(arr.length / 2);
+    const firstQuartileLocation = Math.floor(arr.length / 4);
+    const thirdQuartileLocation = 3 * Math.floor(arr.length / 4);
```
Should use `//` divide?

ilan-gold replied:
I think this just creates a comment, unless there is some other way to use it.
